File: stepik-union/src/main/java/org/stepik/core/courseFormat/stepHelpers/TableQuizHelper.java
package org.stepik.core.courseFormat.stepHelpers;

import com.intellij.openapi.project.Project;
import org.jetbrains.annotations.NotNull;
import org.stepik.api.objects.submissions.Choice;
import org.stepik.api.objects.submissions.Column;
import org.stepik.core.courseFormat.StepNode;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * @author meanmail
 */
public class TableQuizHelper extends QuizHelper {
    private Map<String, Map<String, Boolean>> choices;

    public TableQuizHelper(@NotNull Project project, @NotNull StepNode stepNode) {
        super(project, stepNode);
    }

    @Override
    protected void done() {
        List<Choice> tableChoices = reply.getTableChoices();
        choices = new HashMap<>();

        if (tableChoices.isEmpty()) {
            List<String> rows = getDataset().getRows();
            List<String> cols = getDataset().getColumns();
            for (String row : rows) {
                Map<String, Boolean> map = new HashMap<>();
                for (String column : cols) {
                    map.put(column, false);
                }
                choices.put(row, map);
            }
            return;
        }

        for (Choice choice : tableChoices) {
            HashMap<String, Boolean> map = new HashMap<>();
            for (Column column : choice.getColumns()) {
                map.put(column.getName(), column.getAnswer());
            }
            choices.put(choice.getNameRow(), map);
        }
    }

    @Override
    void fail() {
        choices = new HashMap<>();
    }

    public boolean getChoice(@NotNull String rowName, @NotNull String colName) {
        initStepOptions();
        return choices.getOrDefault(rowName, new HashMap<>()).getOrDefault(colName, false);
    }

    @NotNull
    public String getDescription() {
        initStepOptions();
        return getDataset().getDescription();
    }

    @NotNull
    public List<String> getRows() {
        initStepOptions();
        return getDataset().getRows();
    }

    @NotNull
    public List<String> getColumns() {
        initStepOptions();
        return getDataset().getColumns();
    }

    public boolean isCheckbox() {
        initStepOptions();
        return getDataset().isCheckbox();
    }
}

File: stepik-union/src/main/java/org/stepik/core/courseFormat/stepHelpers/TableQuizHelper.kt
package org.stepik.core.courseFormat.stepHelpers

import com.intellij.openapi.project.Project
import org.stepik.core.courseFormat.StepNode
import java.util.*


class TableQuizHelper(project: Project, stepNode: StepNode) : QuizHelper(project, stepNode) {
    private var choices: MutableMap<String, Map<String, Boolean>>? = null

    val description: String
        get() {
            initStepOptions()
            return dataset.description
        }

    val rows: List<String>
        get() {
            initStepOptions()
            return dataset.rows
        }

    val columns: List<String>
        get() {
            initStepOptions()
            return dataset.columns
        }

    val isCheckbox: Boolean
        get() {
            initStepOptions()
            return dataset.isCheckbox
        }

    override fun done() {
        val tableChoices = reply.tableChoices
        choices = HashMap()

        if (tableChoices.isEmpty()) {
            val rows = dataset.rows
            val cols = dataset.columns
            for (row in rows) {
                val map = HashMap<String, Boolean>()
                for (column in cols) {
                    map[column] = false
                }
                choices!![row] = map
            }
            return
        }

        for (choice in tableChoices) {
            val map = HashMap<String, Boolean>()
            for (column in choice.columns) {
                map[column.name] = column.answer
            }
            choices!![choice.nameRow] = map
        }
    }

    override fun fail() {
        choices = HashMap()
    }

    fun getChoice(rowName: String, colName: String): Boolean {
        initStepOptions()
        return choices!!.getOrDefault(rowName, HashMap()).getOrDefault(colName, false)
    }
}
